grub-core/commands/search.c

```diff
@@ -87,11 +89,40 @@ iterate_device (const char *name, void *data)
     }
 
   /* Skip it if it's not the root device when requested. */
-  root_dev = grub_env_get ("root");
-  if (ctx->flags & SEARCH_FLAGS_ROOTDEV_ONLY &&
-      (name[0] != root_dev[0] || name[1] != root_dev[1] ||
-       name[2] != root_dev[2]))
-    return 0;
+  if (ctx->flags & SEARCH_FLAGS_ROOTDEV_ONLY)
+    {
+      const char *root_dev;
+      root_disk = grub_malloc (grub_strlen(root_dev));
+      if (root_disk == NULL)
+        goto fail;
+      name_disk = grub_malloc (grub_strlen(name));
+      if (name_disk == NULL)
+        goto fail;
+      rest = grub_malloc(grub_strlen(name));
+      if (rest == NULL)
+        goto fail;
+
+      root_dev = grub_env_get ("root");
```

Comment:
This has to be moved before you grub_malloc root_disk (line 95) as you're doing grub_strlen(root_dev) for the malloc size.